### PR TITLE
Rearange documentation prioritizing more useful stuff higher

### PR DESCRIPTION
**Pull request by Indieberrie**
I was kinda confused first time reading this when I figured out how to run test, but still had no clue how to run compiler itself.
I guess it makes more sense that tests are at the bottom, while description of project and setting up should be higher (even than supported features)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,11 +2,45 @@
 
 A compiler from Python to Mindustry's assembly (logic programming language).
 
-To run the tests, simply use the following command:
+The language **is** Python, so all simple programs you can imagine are possible. Check the [supported features][features]. There also might be some missing, so consider [contributing].
+
+Beyond that, the
+supported "system calls" are documented in the [`pyndustri.pyi`] file.
```
Comment (Lonami) [inline README.md:8]:
```suggestion
The language **is** Python, so all simple programs you can imagine are possible.
Check the [supported features][features] to learn what is currently possible.
If you find something is missing, you're welcome to [contribute]!

Beyond that, the supported "system calls" are documented in the [`pyndustri.pyi`] file.
```

File: README.md [hunk 2]
```diff
@@ -213,9 +206,21 @@ Contributors are more than welcome! Maybe you can improve the documentation, add
...
+```
+
 [f-string]: https://docs.python.org/3/reference/lexical_analysis.html#f-strings
 [ip-not-reset]: https://github.com/Anuken/Mindustry/issues/4189
 [limit-k]: https://github.com/Anuken/Mindustry/blob/ab19e6ffbd7a64117cd70d3e3b88806c13822c94/core/src/mindustry/logic/LExecutor.java#L28
 [`pyndustri.pyi`]: pyndustri.pyi
 [`constants.py`]: pyndustric/constants.py
 [`pyndustric`]: pyndustric
+[features]: #Supported-features
+[contributing]: #Contributing
```
Comment (Lonami) [inline README.md:225]:
```suggestion
[contribute]: #Contributing
```

File: README.md [hunk 3]
```diff
@@ -2,11 +2,45 @@
...
+
+## Getting up and running
+
+To install pyndustric as module run:
 
 ```sh
-$ pip install -r dev-requirements.txt
-$ pytest
+$ git clone https://github.com/Lonami/pyndustric.git
+$ pip install -e pyndustric
```
Comment (Lonami) [inline README.md:19]: By the way I'm pretty sure it's possible to `pip install https://github.com/Lonami/pyndustric/archive/refs/heads/master.zip` directly, which does not even require `git` to be installed.

Also curious, you seem to use `install -e`, which is for "editable" installs.

```
  -e, --editable <path/url>   Install a project in editable mode (i.e. setuptools "develop mode") from a local project
                              path or a VCS url.
```

I don't think users who don't want to develop would want to do this, but maybe you had a reason for it.

File: README.md [hunk 4]
```diff
@@ -2,11 +2,45 @@
...
+```
+
+Alternativly, you can run it as module without installation by placing your program file at the root of the project directory.
```
Comment (Lonami) [inline README.md:22]:
```suggestion
Alternatively, you can run it as module without installation by placing your program file at the root of the project directory.
```

**Review by Lonami — CHANGES_REQUESTED** [review]

**Comment by Lonami** [comment]
P.S. super happy to see you're this engaged with this little project ;)

File: README.md (hunk 3 above)
Comment (Indieberrie) [reply README.md]: Reason I did it was because I just copied what I did not thinking much about it. Force of habit
I changed docs now

**Comment by Indieberrie** [comment]
You said it just as I was thinking about doing something else, since I have a lot of stuff on my schedule (instead of no lifing mindustry)
I'll probably contribute much more (not having arrays (lists?) is going on my nerves), but I have no clue when will that be

**Comment by Lonami** [comment]
> not having arrays (lists?) is going on my nerves

Those are tricky because they're no longer a "simple value" which is held in a single variable.

While `for x in [a, b, c]:` could be special-cased to run the loop body with `x` going through all of `a, b, c`, things get more difficult if you were to support to assign the list to a variable, which can then be moved around, passed as parameters, iterated over, etc. Because the type would have to be tracked to properly work on that variable (e.g. in `x = y`, the assignment would be very different for lists and simple values).

Unless… perhaps "list" variables point to their length, and when used as a list, they "just" work. Imagine

```python
x = [1, 2, 3]
for y in x:
  print(y)
```

Could generate (pseudo)

```
x = 3
x__list0 = 1
x__list1 = 2
x__list2 = 3
set loopcounter 0
:loop
jmp loopcounter == 3 to :exitloop
set y x__list{loopcounter}
print y
printflush
jmp :loop
:exitloop
```

Yeah that's some nice food for thought.

**Comment by Lonami** [comment]
(Now as for what would happen if you had `x = 3; for y in x`, well, it would do the same, but it would pick out null every time, kinda "undefined behaviour".)

**Comment by Indieberrie** [comment]
Have you thought about using heap to allocate arrays?
Allocating arrays on the stack (or using infinite registers) is only good when the array size is known at compile time.
As I see, whole idea of python arrays (I use array and list interchangably) is that they are all runtime resizable which means implementing them on stack is a bad idea

**Comment by Lonami** [comment]
The heap (memory cells) can only store numbers (so, no strings, or `@variable`s):

```
set result "hello"
write result cell1 0
read result cell1 0
print result
printflush message1
```

This prints 1. Removing the `read` does print `"hello"`.

It's true, as you point out, infinite variables would not work with dynamic array sizes (since there's no way to refer to a dynamic variable name in mlog).

However, tuples are immutable, so the fixed size makes more sense. Maybe lists could only support numbers (implemented via `Mem`), and tuples would support anything (implemented via infinite registers). But what memory cell would be used for the list?

```python
foo = [1, 2, 3] @ cell1
```

could be made work. Ah, so many ideas!

**Comment by Lonami** [comment]
Let's move this discussion to a separate issue if you have further comments.

**Pull request merged**